### PR TITLE
Add cgo implementation for llama.cpp

### PR DESCRIPTION
**Pull request by dhiltgen**
This change revamps the way ollama wires up llama.cpp for gguf to link directly via cgo instead
of running a subprocess.  Within llama.cpp, a thin facade has been added to server.cpp (via included patch) 
to enable extern "C" access to the main logic to minimize changes to the existing LLM interface.

Mac, Linux, and Windows are supported and manually tested.  

Carries #1268 and #814

### CODE REVIEW

File: llm/llama.cpp/generate_darwin_amd64.go
Comment (jmorganca) [inline llm/llama.cpp/generate_darwin_amd64.go:18]: It would be awesome to remove the need for `go:generate` altogether. We'd probably need to:

1. Vendor vs submodule the c/c++ code and headers
2. There may be a tiny .o or .a file we need to vendor from building the `.cuda` file with `nvcc` (`cmake` does this today).
3. Figure out how to apply any patches (seems like it's generally new code?)

This could make `go build .` just work without the need for cmake or nvcc. It should be even easier on mac (no .cuda file to compile out of band).

This can be a follow-on PR

File: llm/ext_server.go
Comment (BruceMacD) [inline llm/ext_server.go:63]: This allows the request cancellation to be relayed to llama.cpp to stop prediction correct?

File: llm/ext_server.go
Comment (BruceMacD) [inline llm/ext_server.go:73]: Doesn't have to be part of this change specifically, but it might be simpler to call the llama.cpp functions which do predictions and embeddings directly from our cgo rather than sending them through the server logic. That way when we do concurrent requests we don't have 2 layers of request management.

File: llm/ext_server.go
Comment (BruceMacD) [inline llm/ext_server.go:381]: The main thing to close here before was the external subprocess, so that shouldn't be a problem now. Just releasing whatever memory 👍

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go:3]: I'll clean that up before I drop the draft status - personal breadcumb to get local debugging working while I was hacking away at things.

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go:19]: Once I find the right incantations across all the supported OSes and architectures and can identify the common patterns, I'll make sure to distill those into shared ones so this is DRY.

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go]: Turns out I was being too fancy for Windows+cgo and this pattern blew up.  I've refactored the code to not have C callback into Go during the completion flow.

File: llm/ext_server.go
Comment (65a) [inline llm/ext_server.go:23]: I assume it should be possible to put the accelerator-specific flags into separate tag controlled files if Ollama wants to support the other accelerators supported upstream, e.g. ROCm, MKL/OneDNN. Trying to figure out how something like #814 works in this approach.

Comment (dhiltgen) [reply llm/ext_server.go:23]: I'm not sure yet what the best approach is, but once I get things in decent shape with Nvidia, I was planning to explore options on adding #814 into the mix.

File: llm/gpu_darwin.go
Comment (jmorganca) [inline llm/gpu_darwin.go:11]: I think this comment is right – there's till a few crashes folks will run into if they don't have enough free memory (and it doesn't seem like memory paging works).

File: llm/ext_server.go
Comment (jmorganca) [inline llm/ext_server.go:140]: There's a `main_gpu` parameter that's been added since you opened this PR https://github.com/jmorganca/ollama/blob/37d95157dfed0726d751d3a694e9d6cd3c923097/llm/llama.go#L346

File: llm/gpu_darwin.go
Comment (jmorganca) [reply llm/gpu_darwin.go:11]: (Can solve this in a follow up)

File: llm/llama.cpp/patches/0001-Expose-callable-API-for-server.patch
Comment (jmorganca) [inline llm/llama.cpp/patches/0001-Expose-callable-API-for-server.patch:1]: Is there a way to just have this code in the repo (vs as a patch?) Not a huge deal

File: llm/ext_server.go
Comment (jmorganca) [inline llm/ext_server.go:1]: `ggml` models would no longer run with this change. We probably want to send back an error here that notifies the user if we detect a `ggml` model: https://github.com/jmorganca/ollama/blob/main/llm/llm.go#L79

Here's a good error message: https://github.com/jmorganca/ollama/blob/37d95157dfed0726d751d3a694e9d6cd3c923097/server/routes.go#L106C5-L106C21

**Comment by jmorganca** [comment]
This is looking really good. Left a few comments on handling older GGML files

**Review by skydiver1980 — APPROVED** [review]

**Comment by jmorganca** [comment]
There seems to be a `llama.log` that gets dropped in the CWD. It would be great to disable this (I think there's a `LLAMA_` include directive that can be used)

File: llm/llm.go
Comment (jmorganca) [inline llm/llm.go:92]: We can remove all of the `chooseRunners` logic now – this should simplify things quite a bit!

Comment (jmorganca) [reply llm/llm.go:92]: Nevermind, this PR already does this!

File: llm/llm.go
Comment (jmorganca) [inline llm/llm.go:95]: Keep an eye on https://github.com/jmorganca/ollama/pull/1268 for the experience deprecating the ggml runner.

cc @BruceMacD – not sure if you guys want merge these two branches or keep them separate? either works

Comment (BruceMacD) [reply llm/llm.go:95]: I think we should merge them separately otherwise this will be the mother of all changes, but I also don't want to railroad this change. I'll co-ordinate with Daniel.

Comment (dhiltgen) [reply llm/llm.go:95]: Let's get @BruceMacD's change in first, and I'll rebase on top of that, and pick up the MainGPU opt as well.

File: llm/ext_server.go
Comment (mxyng) [inline llm/ext_server.go:9]: This should define `GGML_METAL_NDEBUG` before merging

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go:72]: Can you elaborate?  I'm not quite following the suggestion.

In my manual testing, it seemed like https://github.com/jmorganca/ollama/blob/main/server/routes.go#L94 was sufficient to ensure only one model ran concurrently, and things didn't break, but when I tried to create an integration test and "force" concurrency, things broke in server.cpp either with no slots available, or the contexts getting swapped.  To safeguard for now I added this mutex until I can try to get concurrency sorted out.

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go:99]: I was replicating https://github.com/jmorganca/ollama/blob/main/llm/llama.go#L351-L353 
I'll hard code it to zero to always derive from the model.

**Comment by dhiltgen** [comment]
Comments addressed, rebased, and tested on mac, linux(amd64), and windows.

This will require another rebase after @BruceMacD's change goes in.

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go]: I've updated this to carry #814 and adapt it to the new cgo model.

File: llm/ext_server.go
Comment (BruceMacD) [inline llm/ext_server.go:197]: Can we just set these directly rather than initializing to false and checking the conditional?
```suggestion
	sparams.memory_f16 = C.bool(opts.F16KV)
	sparams.use_mlock = C.bool(opts.UserMLock)
	sparams.use_mmap = C.bool(opts.UserMMap)
	sparams.numa = C.bool(opts.UseNUMA)
```

**Review by BruceMacD — COMMENTED** [review]
I'm really liking this after trying it today, noticed a couple things

File: llm/ext_server.go
Comment (BruceMacD) [reply llm/ext_server.go:89]: Just a heads up, locking here might not guarantee availability as the slots are freed in a separate thread from the  predictions.

**Comment by dhiltgen** [comment]
Multimodal is working now on this branch.  It seems llama.cpp `b1632` doesn't work, but reverting back to `b1627` gets image processing working.

**Comment by dhiltgen** [comment]
Now buildable on a linux system without CUDA lib's installed to produce a CPU only build.

File: llm/ext_server.go
Comment (65a) [reply llm/ext_server.go]: Nice, let me know if you want to me test it, but appreciate you carrying the change!

File: gpu/gpu_info_cuda.c
Comment (BruceMacD) [inline gpu/gpu_info_cuda.c:31]: I think this might break `wsl` which a lot of people are currently using to run Ollama on Windows. The GPU is accessible to WSL (I can see it by running `nvidia-smi`), but I see the error `CUDA not detected: Unable to load libnvidia-ml.so library to query for nvidia GPUs: ... : no such file or directory` in the logs. This file on the host system wont be accessible in WSL but the GPU will still work.

Comment (BruceMacD) [reply gpu/gpu_info_cuda.c:31]: I think using `nvidia-smi` as a fallback could be a good solution.

File: llm/ext_server.go
Comment (jmorganca) [inline llm/ext_server.go:70]: more of a nit: do we need an interface here? the `LLM` interface as well seems like overkill

Comment (jmorganca) [reply llm/ext_server.go:70]: Ah, just realized this is for the rocm shim! Might be worth nuking the LLM interface in a follow up at some point if it's mostly an added layer of complexity

**Comment by jmorganca** [comment]
Looking really solid – will test this out some more this evening!

**Comment by dhiltgen** [comment]
Rebased and tested on mac, linux (cuda + rocm)

I added support recently to be able to build without GPU compiled in, but missed a corner case where on radeon systems, the cpu build fails with `Error: ROCm shim library not included in this build of ollama. Radeon GPUs are not supported`.  The GPU enabled build works fine.  I'll investigate that and either push another commit to fix it, or we can follow up with a subsequent PR later.

**Comment by dhiltgen** [comment]
The CPU + radeon detected corner case is mitigated.  It's running slower than it should, so I think there's probably more refinement we can do, but it runs now.

File: gpu/gpu_info_cuda.c
Comment (dhiltgen) [reply gpu/gpu_info_cuda.c:32]: Hmm... that's odd.  I would have thought that if nvidia-smi was inside the WSL vhd image that it would also have the underlying library that CLI depends on, which is what I'm tying to load.  Maybe we do have to fallback to CLI, but let me poke around a bit and see if there's another way...

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go:70]: Yeah, I had this same thought after you brought this up a few days ago...   My only slight hesitation was it might be helpful for unit testing, but I'm not sure.  I'd suggest maybe we leave it for the moment and flesh out some more test cases, and if it's not helpful on that front, ~combine LLM and extServer interfaces and simplify them down to what's actually necessary.

File: gpu/gpu_info_cuda.c
Comment (dhiltgen) [reply gpu/gpu_info_cuda.c:32]: I've tried a couple possibilities, but no solution yet...

Turns out there's a `/usr/local/cuda/lib64/stubs/libnvidia-ml.so` in WSL2 if you have the CUDA libs installed, but at runtime that fails with a wall of text error and code:
```
NVML_ERROR_DRIVER_NOT_LOADED = 9
NVIDIA driver is not loaded.
```

I also found `/mnt/c/Windows/System32/lxss/lib/libnvidia-ml.so.1` is present, however that also fails at runtime with an init error of:
```
NVML_ERROR_OPERATING_SYSTEM = 17
The GPU control device has been blocked by the operating system/cgroups.
```
(root or normal user doesn't matter)

That last error feels like it might be solvable if we can figure out how things are wired up, but I'm not sure.

My inclination would be to merge what we've got and note this as a limitation on WSL2 for the pre-release (Your call @jmorganca ). I imagine a lot of people running in WSL2 will be happier with it running natively, which it does in its current form.  We can continue to investigate if there's a viable solution to get libnvidia-ml.so to load inside WSL2, and if not, fallback to the CLI as @BruceMacD suggested.

Comment (dhiltgen) [reply gpu/gpu_info_cuda.c:33]: Scratch that - I mistakenly tested under wsl1 which won't work.  Reset my setup to wsl2 and found a fix.

There's probably room to improve detection and error reporting for misconfigurations.

**Comment by jmorganca** [comment]
I think we'll need to embed `ggml-metal.metal` and spit it out in the cwd, if it doesn't yet exist (or is different than what's bundled). For the Mac app we can add it to the Ollama.app build (here's the change that took it out we'd need to revert: https://github.com/jmorganca/ollama/commit/a5dbcf2e735b8c87bc9919588e99ff4a814a09da)

Here's the code that used to do that for the Go side:

* `go:embed`: https://github.com/jmorganca/ollama/blob/f4432e1dbac6ca8af83979bd324cc0483fc8db7a/llm/llama.go#L106C17-L106C17
* Writing it to disk: https://github.com/jmorganca/ollama/blob/f4432e1dbac6ca8af83979bd324cc0483fc8db7a/llm/llama_darwin.go#L31

File: README.md
Comment (BruceMacD) [inline README.md:203]: I think we should move these extended docs down to `docs/development.md` and link to them from the main README

File: gpu/gpu.go
Comment (BruceMacD) [inline gpu/gpu.go:129]: We saw issues on smaller GPUs when trying to load big models, while falling back to CPU allows the model to run (it can be paged out of memory), so if this still works the same way might be worth keeping the behavior.

File: scripts/build_remote.py
Comment (BruceMacD) [inline scripts/build_remote.py:1]: Is this used anywhere? If not maybe we should make this a follow up

File: Dockerfile.cpu
Comment (dhiltgen) [reply Dockerfile.cpu:1]: The CUDA based built-in should be able to fallback to CPU, so we don't have to ship this CPU-only build as part of the official release.  However, during development of this cgo change, I wound up breaking CPU-only builds for a while, and added this Dockerfile to be able to quickly/easily perform a linux build without CUDA or ROCm present.  I think what likely makes sense is to include some sort of CPU-only build in our CI to ensure we don't regress on that for people who checkout the repo and want to build from source for CPU only.   Depending on how we choose to wire up github actions, we might leverage this.

File: scripts/build_remote.py
Comment (dhiltgen) [reply scripts/build_remote.py:1]: I've been using this to build on windows and it's been helpful for iterative dev/test cycles.

**Comment by dhiltgen** [comment]
While it should be possible to build on older distros from source, the containerized build for official images does narrow our support matrix due to [ROCm compatibility requirements](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/tutorial/quick-start.html#).  As currently set up, here's the minimum OS versions the binaries should support

| **OS** | **Minimum Version**|
|------|--------------------|
| Ubuntu | 20.04 |
| Debian | 11 |
| Rocky | 9 |
| Fedora | 32 |
| SLES | 15-SP5 |

File: llm/shim_ext_server.go
Comment (BruceMacD) [inline llm/shim_ext_server.go:126]: Should we just fallback to CPU in this case?

File: server/images.go
Comment (BruceMacD) [inline server/images.go:422]: note to reader: this is also part of the ggml deprecation

**Review by BruceMacD — APPROVED** [review]
I'm feeling pretty good about getting this in now that v0.1.17 is out, just need blessing from some other members of the team and I say ship it.

File: llm/shim_ext_server.go
Comment (dhiltgen) [reply llm/shim_ext_server.go:126]: My thinking was if you've got all the things set up properly except the permissions, it may be better to hard fail so it's "obvious" instead of run on CPU slowly since we don't have a way to expose that out to the CLI at present.

File: docs/modelfile.md
Comment (mxyng) [inline docs/modelfile.md:191]: GGML is still the right term. GGUF is a specific type of GGML which only applies to models, not LoRA. LoRA files in GGML file format are actually type GGLA

File: gpu/gpu.go
Comment (mxyng) [inline gpu/gpu.go:36]: It would make things considerably easier IMO if errors were well defined and `*_init` returned an int error. e.g.

```go
var cuda C.cuda_init_resp_t
if retval := C.cuda_init(&cuda); retval == 0 {
  gpuHandles.cuda = &cuda.ch
  return
}

var rocm C.rocm_init_resp_t
if retval := C.rocm_init(&rocm); retval == 0 {
  gpuHandles.rocm = &rocm.rh
  return
}

// cpu
```

File: llm/ext_server.go
Comment (jmorganca) [inline llm/ext_server.go:5]: I think we want to set `-DSERVER_VERBOSE=0` since it's currently outputting each token in the logs

**Comment by jmorganca** [comment]
Tested this on CUDA, previous issues seem gone, which is awesome!

Added an inline comment RE disabling the verbose logging

File: Dockerfile.build
Comment (jmorganca) [inline Dockerfile.build:56]: Do we want to download and build CLBlast on arm64 as well?

File: llm/llama.cpp/gen_linux.sh
Comment (BruceMacD) [inline llm/llama.cpp/gen_linux.sh:29]: this `git_module_setup` step will have to be done before creating the `OLLAMA_DYN_LIB_DIR` or the submodule clone will fail on the first generation due to the directory not being empty:
```
go generate ./...
+ set -o pipefail
+ echo 'Starting linux generate script'
Starting linux generate script
+ '[' -z '' -a -x /usr/local/cuda/bin/nvcc ']'
+ COMMON_CMAKE_DEFS='-DCMAKE_POSITION_INDEPENDENT_CODE=on -DLLAMA_ACCELERATE=on -DLLAMA_NATIVE=off -DLLAMA_AVX=on -DLLAMA_AVX2=off -DLLAMA_AVX512=off -DLLAMA_FMA=off -DLLAMA_F16C=off'
+ OLLAMA_DYN_LIB_DIR=gguf/build/lib
+ mkdir -p gguf/build/lib
+ touch gguf/build/lib/.generated
++ dirname ./gen_linux.sh
+ source ./gen_common.sh
+ init_vars
+ LLAMACPP_DIR=gguf
+ PATCHES=0001-Expose-callable-API-for-server.patch
+ CMAKE_DEFS=-DLLAMA_ACCELERATE=on
+ CMAKE_TARGETS='--target ggml --target ggml_static --target llama --target build_info --target common --target ext_server --target llava_static'
+ echo ''
+ grep -- -g
+ CMAKE_DEFS='-DCMAKE_BUILD_TYPE=Release -DLLAMA_ACCELERATE=on'
+ git_module_setup
+ '[' -n '' ']'
+ git submodule init
+ git submodule update --force gguf
fatal: destination path '/home/bruce/dhiltgen/ollama/llm/llama.cpp/gguf' already exists and is not an empty directory.
fatal: clone of 'https://github.com/ggerganov/llama.cpp.git' into submodule path '/home/bruce/dhiltgen/ollama/llm/llama.cpp/gguf' failed
Failed to clone 'llm/llama.cpp/gguf'. Retry scheduled
fatal: destination path '/home/bruce/dhiltgen/ollama/llm/llama.cpp/gguf' already exists and is not an empty directory.
fatal: clone of 'https://github.com/ggerganov/llama.cpp.git' into submodule path '/home/bruce/dhiltgen/ollama/llm/llama.cpp/gguf' failed
Failed to clone 'llm/llama.cpp/gguf' a second time, aborting
llm/llama.cpp/generate_linux.go:3: running "bash": exit status 1
```

File: llm/ext_server.go
Comment (dhiltgen) [reply llm/ext_server.go:5]: I'll follow up with a secondary PR to quiet down the server verbosity.

File: gpu/gpu.go
Comment (dhiltgen) [reply gpu/gpu.go:36]: I'll take a look at tidying up the error handling in a follow up PR.

**Comment by tecosaur** [comment]
I see in `docs/development.md` it's stated what's required to build with support for AMD GPUs. I presume the release binaries are built with this though, making me wonder if we could get similar clarity on what needs to be installed at runtime?

**Comment by tecosaur** [comment]
I'm still interested in hearing what's required for AMD GPUs at runtime, if anybody happens to know :slightly_smiling_face:

**Comment by dhiltgen** [comment]
@tecosaur we're still working on some rough edges to improve the experience with AMD GPUs.  Once those are addressed, we'll update the docs and install script accordingly.  Until then, try to install the ROCm v6 library and corresponding driver https://rocm.docs.amd.com/projects/install-on-linux/en/latest/tutorial/quick-start.html

**Comment by tecosaur** [comment]
Ah cool. Not seeing anything on this I was getting worried this aspect had been overlooked. I've just managed to get the "standard set" of ROCm packages installed, but I see some places seem to say more are needed depending on the application (e.g. https://www.reddit.com/r/LocalLLaMA/comments/170tghx/guide_installing_rocmhip_for_llamacpp_on_linux/).

For what it's worth, I'm happy to do a little testing and write up what does/doesn't work for me :slightly_smiling_face:.

**Comment by tecosaur** [comment]
Ah, I see that #738 might be a better place for me to go.

**Comment by dhiltgen** [comment]
> install the ROCm v6 library 

Oops, I meant v5 library. (that would be one of the rough edges we're working on)

**Comment by tecosaur** [comment]
You can see me hitting that rough edge in #738 :stuck_out_tongue:

**Pull request merged**
